src/app/InteractionModelEngine.cpp

```diff
@@ -462,40 +397,96 @@ void InteractionModelEngine::OnResponseTimeout(Messaging::ExchangeContext * ec)
                     ChipLogValueExchange(ec));
 }
 
-CHIP_ERROR InteractionModelEngine::SendReadRequest(ReadPrepareParams & aReadPrepareParams, ReadClient::Callback * aCallback)
+uint16_t InteractionModelEngine::GetWriteClientArrayIndex(const WriteClient * const apWriteClient) const
 {
-    ReadClient * client = nullptr;
-    CHIP_ERROR err      = CHIP_NO_ERROR;
-    ReturnErrorOnFailure(NewReadClient(&client, ReadClient::InteractionType::Read, aCallback));
-    err = client->SendRequest(aReadPrepareParams);
-    if (err != CHIP_NO_ERROR)
-    {
-        client->Shutdown();
-    }
-    return err;
+    return static_cast<uint16_t>(apWriteClient - mWriteClients);
 }
 
-CHIP_ERROR InteractionModelEngine::SendSubscribeRequest(ReadPrepareParams & aReadPrepareParams, ReadClient::Callback * aCallback)
+uint16_t InteractionModelEngine::GetReadHandlerArrayIndex(const ReadHandler * const apReadHandler) const
 {
-    ReadClient * client = nullptr;
-    ReturnErrorOnFailure(NewReadClient(&client, ReadClient::InteractionType::Subscribe, aCallback));
-    ReturnErrorOnFailure(client->SendRequest(aReadPrepareParams));
-    return CHIP_NO_ERROR;
+    return static_cast<uint16_t>(apReadHandler - mReadHandlers);
 }
 
-uint16_t InteractionModelEngine::GetReadClientArrayIndex(const ReadClient * const apReadClient) const
+void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
 {
-    return static_cast<uint16_t>(apReadClient - mReadClients);
+    ReadClient * pPrevListItem = nullptr;
+    ReadClient * pCurListItem  = mpActiveReadClientList;
+
+    while (pCurListItem)
+    {
+        pPrevListItem = pCurListItem;
+        pCurListItem  = pCurListItem->GetNextClient();
+    }
+
+    if (pPrevListItem)
+    {
+        pPrevListItem->SetNextClient(apReadClient);
+    }
+    else
+    {
+        mpActiveReadClientList = apReadClient;
+    }
+
+    apReadClient->SetNextClient(nullptr);
 }
 
-uint16_t InteractionModelEngine::GetWriteClientArrayIndex(const WriteClient * const apWriteClient) const
+void InteractionModelEngine::RemoveReadClient(ReadClient * apReadClient)
 {
-    return static_cast<uint16_t>(apWriteClient - mWriteClients);
+    ReadClient * pPrevListItem = nullptr;
+    ReadClient * pCurListItem  = mpActiveReadClientList;
+
+    while (pCurListItem != apReadClient)
+    {
+        pPrevListItem = pCurListItem;
+        pCurListItem  = pCurListItem->GetNextClient();
+    }
+
+    //
+    // Item must exist in this tracker list. If not, there's a bug somewhere.
+    //
+    VerifyOrDie(pCurListItem != nullptr);
```

Comment:
I'm not sure that's true.  Say someone shuts down the IM engine, we clear the tracker list, then _after_ that the ReadClient is destroyed.  We'll crash on this line, no?

Reply:
Actually, if the IM engine is shut-down, _technically,_ you shouldn't be calling any methods on it till it is re-initialized right? In which case, the weak-reference model won't quite work right?

Reply:
Given that we are storing a pointer to the IM engine in the ReadClient, as opposed to it directly grabbing the singleton, one sane behavior might be:

1. Construction of a ReadClient requires a pointer to the IM engine.  It's on you to ensure this is a live IM engine.  Constructor registers the ReadClient.
2. Destruction of a ReadClient null-checks mpImEngine and if non-null unregisters.
3. Shutdown of IM engine walks the registered list and nulls out the pointer to itself on all of them (or calls some function that nulls out both that and the "next" pointer, perhaps?), then nulls out mpActiveReadClientList.

I think given those pieces we could keep the VerifyOrDie in `RemoveReadClient`, because then it really would be something very much unexpected to not have the client int the list.